Sources/TensorFlow/LazyTensor/LazyTensorOperation.swift

```diff
@@ -0,0 +1,551 @@
+
+/// The `TF_Tensor *` type.
+typealias CTFTensor = OpaquePointer
+
+public class LazyTensor : _AnyTensorHandle {
+    enum Handle {
+        /// Bool indicates if this concrete TFETensorhandle was a result of
+        /// materialization.
+       case conc(TFETensorHandle, Bool)
+        /// Bool indicates whether this is held by a swift variable,
+        /// i.e., is this a live tensor?
+       case sym(LazyTensorOperation, Int, Bool)
+    }
+
+    let handle: Handle
+
+    public var _tfeTensorHandle: TFETensorHandle {
+        switch (handle) {
+        case Handle.conc(let h, _):
+            return h
+        case Handle.sym(let op, let index, _):
+            assert(false, "TODO: to be send out in a separate PR.")
+            // return op.materialized(index: index)
+        }
+    }
+
+    init(_ base: TFETensorHandle) {
+        handle = Handle.conc(base, false)
+    }
+
+    init(_materialized base: TFETensorHandle) {
+        handle = Handle.conc(base, true)
+    }
+
+    init(_lazy op: LazyTensorOperation, index: Int) {
+        precondition(index < op.outputCount, "Symbolic Tensor Index is out-of-bounds")
+        handle = Handle.sym(op, index, false)
+    }
+
+    init(_lazyLive op: LazyTensorOperation, index: Int) {
+        precondition(index < op.outputCount, "Symbolic Tensor Index is out-of-bounds")
+        handle = Handle.sym(op, index, true)
+    }
+}
+
+public class LazyTensorOperation : TensorOperation {
+    public typealias TensorValueHandle = LazyTensor
+
+    enum Input {
+        case single(LazyTensor)
+        case list([LazyTensor])
+    }
+
+    enum Attribute {
+        case BoolValue(Bool)
+        case IntValue(Int)
+        case FloatValue(Float)
+        case DoubleValue(Double)
+        case StringValue(String)
+        case BoolArray([Bool])
+        case IntArray([Int])
+        case FloatArray([Float])
+        case DoubleArray([Double])
+        case StringArray([String])
+        case ConstTensor(TFETensorHandle)
+        case TensorDataTypeValue(TensorDataType)
+    }
+
+    let name: String
+    let outputCount: Int
+    var inputs: [Input]
+    var attrs: [String: Attribute]
+    var outputs: [TFETensorHandle]?
+    var id: String?
+
+    static public var materializationCallback: (String) -> () = {
+        (s: String) in return }
+
+    public static func registerMaterializationCallback(
+        f: @escaping (String) -> ()) {
+        materializationCallback = f
+    }
+
+    public var nameWithID: String {
+        if let id = self.id {
+            return "\(name)_\(id)"
+        } else {
+            return "\(name)_\(ObjectIdentifier(self))"
+        }
+    }
+
+    public init(_withID id: String?, _ name: String, _ outputCount: Int) {
+        self.name = name
+        self.inputs = []
+        self.attrs = [:]
+        self.outputCount = outputCount
+        self.outputs = nil
+        self.id = id
+    }
+
+    public required convenience init(_ name: String, _ outputCount: Int) {
```

Comment:
Compensate for weak type information with argument labels. Also I'm not sure this needs to be a public API.
```suggestion
    required convenience init(_ name: String, _ outputCount: Int) {
```

Reply:
Removed public. I will defer adding the labels as this is part of the `TFTensorOperation` interface, which is also used by `TFE_Op`